### PR TITLE
[stable/vpa] Add configurable value for tests.securityContext

### LINKED ISSUE #1143: Enable setting securityContext  on  test containers
**Is your feature request related to a problem? Please describe.**

We run a strict gatekeeper policy to prevent pods _without_  security context set from being deployed. When we run goldilocks with VPA, we cannot deploy VPA as the test fails due to gate keeper violations.

**Describe the solution you'd like**

I would like a helm value option for `securityContext` that can be set and passed through to the helm test to enable configuration of the test pods.

**Describe alternatives you've considered**

Using a tool to "wedge" these spec changes in at run time

**Additional context**

This is supported in almost every other aspect of the goldilocks and VPA charts, I think continuing this pattern would be very useful.

### PR DESCRIPTION
**Pull request by davidblum**
**Why This PR?**

This PR enables configuration for securityContext in the pods run as helm tests.

Fixes https://github.com/FairwindsOps/charts/issues/1143

**Changes**
Changes proposed in this pull request:

* add configuration for securityContext for vpa/templates/tests

**Checklist:**

* [x] I have included the name of the chart in the title of this PR in square brackets i.e. `[stable/goldilocks]`.
* [x] I have updated the chart version in `Chart.yaml` following Semantic Versioning.
* [x] Any new values are backwards compatible and/or have sensible default.
* [x] Any new values have been added to the README for the Chart, or `helm-docs --sort-values-order=file` has been run for the charts that support it.

### CODE REVIEW

**Review by sudermanjr — APPROVED** [review]
Looks great, thanks! 

Once the tests pass and the CLA is signed we can get this merged.

**Comment by sudermanjr** [comment]
This might be helpful: https://github.com/FairwindsOps/charts#generating-docs

**Comment by davidblum** [comment]
@sudermanjr CLA signed, docs updated. However its failing on a versioning issue. Should i have bumped the chart version in this PR? or does that happen automatically later?

**Comment by sudermanjr** [comment]
The version bump you did looks correct. Not sure what's up with e2e tests.

**Comment by sudermanjr** [comment]
Here's the error:

```
Error: UPGRADE FAILED: template: vpa/templates/tests/create-vpa.yaml:17:26: executing "vpa/templates/tests/create-vpa.yaml" at <.Values.tests.securityContext>: nil pointer evaluating interface {}.securityContext
```

File: stable/vpa/templates/tests/create-vpa.yaml [hunk 1]
```diff
@@ -13,6 +13,8 @@ spec:
   serviceAccountName: {{ include "vpa.fullname" . }}-test
   containers:
     - name: test
+      securityContext:
+        {{- toYaml .Values.tests.securityContext | nindent 8 }}
```
Comment (sudermanjr) [inline stable/vpa/templates/tests/create-vpa.yaml]: I would check the nindent value here. Quick guess is that it should be 6, but I'm not sure. (applies to other files too)

**Comment by davidblum** [comment]
@sudermanjr I've tried a few variations, and your suggestion.

But I can't get this to pass CI. The change does work locally.

Any thoughts?

**Comment by sudermanjr** [comment]
Apologies for losing track of this. Could you please bump the version again and I'll take a closer look?

**Comment by sudermanjr** [comment]
Apologies for losing track of this. When I install with these values and then run `helm test` I am getting:

```
3m ▶ helm test vpa
W0515 11:44:16.313261   92803 warnings.go:70] unknown field "spec.containers[0].allowPrivilegeEscalation"
W0515 11:44:16.313275   92803 warnings.go:70] unknown field "spec.containers[0].capabilities"
W0515 11:44:16.313277   92803 warnings.go:70] unknown field "spec.containers[0].readOnlyRootFilesystem"
W0515 11:44:16.313280   92803 warnings.go:70] unknown field "spec.containers[0].runAsGroup"
W0515 11:44:16.313281   92803 warnings.go:70] unknown field "spec.containers[0].runAsNonRoot"
W0515 11:44:16.313283   92803 warnings.go:70] unknown field "spec.containers[0].runAsUser"
W0515 11:44:19.770536   92803 warnings.go:70] unknown field "spec.containers[0].allowPrivilegeEscalation"
W0515 11:44:19.770560   92803 warnings.go:70] unknown field "spec.containers[0].capabilities"
W0515 11:44:19.770565   92803 warnings.go:70] unknown field "spec.containers[0].readOnlyRootFilesystem"
W0515 11:44:19.770570   92803 warnings.go:70] unknown field "spec.containers[0].runAsGroup"
W0515 11:44:19.770574   92803 warnings.go:70] unknown field "spec.containers[0].runAsNonRoot"
W0515 11:44:19.770578   92803 warnings.go:70] unknown field "spec.containers[0].runAsUser"
```

File: stable/vpa/templates/tests/crd-available.yaml [hunk 2]
```diff
@@ -14,6 +14,8 @@ spec:
   serviceAccountName:  {{ include "vpa.fullname" . }}-test
   containers:
     - name: test
+      securityContext:
+      {{- toYaml .Values.tests.securityContext | nindent 6 }}
```
Comment (sudermanjr) [inline stable/vpa/templates/tests/crd-available.yaml]:
```suggestion
      {{- toYaml .Values.tests.securityContext | nindent 8 }}
```

File: stable/vpa/templates/tests/crd-available.yaml [hunk 3]
```diff
@@ -37,6 +39,8 @@ spec:
   serviceAccountName: {{ include "vpa.fullname" . }}-test
   containers:
     - name: test
+      securityContext:
+      {{- toYaml .Values.tests.securityContext | nindent 6 }}
```
Comment (sudermanjr) [inline stable/vpa/templates/tests/crd-available.yaml]:
```suggestion
      {{- toYaml .Values.tests.securityContext | nindent 8 }}
```

File: stable/vpa/templates/tests/create-vpa.yaml [hunk 4]
```diff
@@ -14,6 +14,8 @@ spec:
   serviceAccountName: {{ include "vpa.fullname" . }}-test
   containers:
     - name: test
+      securityContext:
+      {{- toYaml .Values.tests.securityContext | nindent 6 }}
```
Comment (sudermanjr) [inline stable/vpa/templates/tests/create-vpa.yaml]:
```suggestion
      {{- toYaml .Values.tests.securityContext | nindent 8 }}
```

File: stable/vpa/templates/tests/metrics.yaml [hunk 5]
```diff
@@ -15,6 +15,8 @@ spec:
   serviceAccountName: {{ include "vpa.fullname" . }}-test
   containers:
     - name: test
+      securityContext:
+      {{- toYaml .Values.tests.securityContext | nindent 6 }}
```
Comment (sudermanjr) [inline stable/vpa/templates/tests/metrics.yaml]:
```suggestion
      {{- toYaml .Values.tests.securityContext | nindent 8 }}
```

File: stable/vpa/ci/test-values.yaml [hunk 6]
```diff
@@ -23,3 +23,11 @@ admissionController:
...
   app: test
   foo: bar
+tests:
+  securityContext:
+    runAsUser: 1984
+    runAsGroup: 1984
+    allowPrivilegeEscalation: false
+    capabilities:
+      drop:
+        - "ALL"
```
Comment (sudermanjr) [inline stable/vpa/ci/test-values.yaml]: Since this is set as default in values.yaml, this is actually unnecessary now.
```suggestion
```

**Review by sudermanjr — CHANGES_REQUESTED** [review]

**Comment by sudermanjr** [comment]
I now get no errors locally, but still get this in the CI job. I'm really not sure what's causing that.

```
Error: UPGRADE FAILED: template: vpa/templates/tests/create-vpa.yaml:18:24: executing "vpa/templates/tests/create-vpa.yaml" at <.Values.tests.securityContext>: nil pointer evaluating interface {}.securityContext
```

**Comment by davidblum** [comment]
@sudermanjr This is the exact issue(s) I ran into as well.

I'm not quite sure which is the source of truth, CI or local build? But I'm concerned that they don't line up...

**Comment by sudermanjr** [comment]
My only guess is that the version of helm in our CI is outdated and doing something odd. I will look into it a bit

**Comment by sudermanjr** [comment]
I think I figured it out. The chart testing procedure does a `helm install` of the _previous_ revision, and then it does a `helm upgrade --reuse-values` with the new version in order to ensure backward compatibility of values. So in this case, I think we need to wrap the security context in an `{{- if .Values.tests }}` block in order to make the upgrade/test work.


you can reproduce this now by doing this:

```
# checkout master
helm install vpa stable/vpa --namespace vpa --create-namespace --values ci/test-values.yaml
helm test vpa -n vpa

# checkout your branch
helm upgrade vpa stable/vpa --namespace vpa --reuse-values
helm test vpa -n vpa
```


Working git diff:

```
diff --git a/stable/vpa/templates/tests/crd-available.yaml b/stable/vpa/templates/tests/crd-available.yaml
index 12e2983..bc5e923 100644
--- a/stable/vpa/templates/tests/crd-available.yaml
+++ b/stable/vpa/templates/tests/crd-available.yaml
@@ -14,8 +14,10 @@ spec:
   serviceAccountName:  {{ include "vpa.fullname" . }}-test
   containers:
     - name: test
+      {{- if .Values.tests }}
       securityContext:
       {{- toYaml .Values.tests.securityContext | nindent 8 }}
+      {{- end }}
       image: quay.io/reactiveops/ci-images:v11-alpine
       command: ['kubectl']
       args:
@@ -39,8 +41,10 @@ spec:
   serviceAccountName: {{ include "vpa.fullname" . }}-test
   containers:
     - name: test
+      {{- if .Values.tests }}
       securityContext:
       {{- toYaml .Values.tests.securityContext | nindent 8 }}
+      {{- end }}
       image: quay.io/reactiveops/ci-images:v11-alpine
       command: ['kubectl']
       args:
diff --git a/stable/vpa/templates/tests/create-vpa.yaml b/stable/vpa/templates/tests/create-vpa.yaml
index cb7a2db..701494e 100644
--- a/stable/vpa/templates/tests/create-vpa.yaml
+++ b/stable/vpa/templates/tests/create-vpa.yaml
@@ -14,8 +14,10 @@ spec:
   serviceAccountName: {{ include "vpa.fullname" . }}-test
   containers:
     - name: test
+      {{- if .Values.tests }}
       securityContext:
       {{- toYaml .Values.tests.securityContext | nindent 8 }}
+      {{- end }}
       image: quay.io/reactiveops/ci-images:v11-alpine
       command: ['bash']
       args:

```

File: stable/vpa/templates/tests/crd-available.yaml [hunk 7]
```diff
@@ -14,6 +14,8 @@ spec:
   serviceAccountName:  {{ include "vpa.fullname" . }}-test
   containers:
     - name: test
+      securityContext:
+      {{- toYaml .Values.tests.securityContext | nindent 8 }}
```
Comment (sudermanjr) [inline stable/vpa/templates/tests/crd-available.yaml:19]:
```suggestion
      {{- if .Values.tests }}
      securityContext:
      {{- toYaml .Values.tests.securityContext | nindent 8 }}
      {{- end }}
```

File: stable/vpa/templates/tests/crd-available.yaml [hunk 8]
```diff
@@ -37,6 +39,8 @@ spec:
   serviceAccountName: {{ include "vpa.fullname" . }}-test
   containers:
     - name: test
+      securityContext:
+      {{- toYaml .Values.tests.securityContext | nindent 8 }}
```
Comment (sudermanjr) [inline stable/vpa/templates/tests/crd-available.yaml:46]:
```suggestion
      {{- if .Values.tests }}
      securityContext:
      {{- toYaml .Values.tests.securityContext | nindent 8 }}
      {{- end }}
```

File: stable/vpa/templates/tests/create-vpa.yaml [hunk 9]
```diff
@@ -14,6 +14,8 @@ spec:
   serviceAccountName: {{ include "vpa.fullname" . }}-test
   containers:
     - name: test
+      securityContext:
+      {{- toYaml .Values.tests.securityContext | nindent 8 }}
```
Comment (sudermanjr) [inline stable/vpa/templates/tests/create-vpa.yaml:19]:
```suggestion
      {{- if .Values.tests }}
      securityContext:
      {{- toYaml .Values.tests.securityContext | nindent 8 }}
      {{- end }}
```

File: stable/vpa/templates/tests/metrics.yaml [hunk 10]
```diff
@@ -15,6 +15,8 @@ spec:
   serviceAccountName: {{ include "vpa.fullname" . }}-test
   containers:
     - name: test
+      securityContext:
+      {{- toYaml .Values.tests.securityContext | nindent 8 }}
```
Comment (sudermanjr) [inline stable/vpa/templates/tests/metrics.yaml:20]:
```suggestion
      {{- if .Values.tests }}
      securityContext:
      {{- toYaml .Values.tests.securityContext | nindent 8 }}
      {{- end }}
```

**Review by sudermanjr — CHANGES_REQUESTED** [review]

**Review by sudermanjr — APPROVED** [review]

**Review by sudermanjr — APPROVED** [review]

**Comment by sudermanjr** [comment]
Thanks again for the contribution, and for staying engaged while we work through this!

**Comment by davidblum** [comment]
@sudermanjr Oh man, that is a sneaky bug! Great find.

Thanks for your help and openness to this issue!

**Pull request merged**